src/combinatorics/additive/salem_spencer.lean

```diff
@@ -0,0 +1,439 @@
+/-
+Copyright (c) 2021 Yaël Dillies, Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Bhavik Mehta
+-/
+import analysis.asymptotics.asymptotics
+
+/-!
+# Salem-Spencer sets and Roth numbers
+
+This file defines Salem-Spencer sets, the Roth number of a set and calculates small Roth numbers.
+
+A Salem-Spencer set is a set without arithmetic progressions of length `3`. Equivalently, the
+average of any two distinct elements is not in the set.
+
+The Roth number of a finset is the size of its biggest Salem-Spencer subset. This is a more general
+definition than the one often found in mathematical litterature, where the `n`-th Roth number is
+the size of the biggest Salem-Spencer subset of `{0, ..., n - 1}`.
+
+## Main declarations
+
+* `mul_salem_spencer`: Predicate for a set to be multiplicative Salem-Spencer.
+* `add_salem_spencer`: Predicate for a set to be Salem-Spencer.
+* `roth_number`: The Roth number of a finset.
+* `roth_number_nat`: The Roth number of a natural. This corresponds to
+  `roth_number (finset.range n)`.
+
+## TODO
+
+Can we calculate small Roth numbers quicker. The current algorithm to decide `roth_number_nat n ≤ m`
+is `O (n.choose m * m^2)`.
+
+## Tags
+
+Salem-Spencer, Roth, arithmetic progression, average
+-/
+
+open finset nat
+
+variables {α β : Type*}
+
+section salem_spencer
+section monoid
+variables [monoid α] [monoid β] (s t : set α)
+
+/-- A multiplicative Salem-Spencer, aka non averaging, set `s` in a monoid is a set such that the
+multiplicative average of any two distinct elements is not in the set. -/
+@[to_additive "A Salem-Spencer, aka non averaging, set `s` in an additive monoid
+is a set such that the average of any two distinct elements is not in the set."]
+def mul_salem_spencer : Prop := ∀ ⦃a b c⦄, a ∈ s → b ∈ s → c ∈ s → a * b = c * c → a = b
+
+/-- Whether a given finset is Salem-Spencer is decidable. -/
+@[to_additive]
+instance {α : Type*} [decidable_eq α] [monoid α] {s : finset α} :
+  decidable (mul_salem_spencer (s : set α)) :=
+decidable_of_iff (∀ a ∈ s, ∀ b ∈ s, ∀ c ∈ s, a * b = c * c → a = b)
+  ⟨λ h a b c ha hb hc, h a ha b hb c hc, λ h a ha b hb c hc, h ha hb hc⟩
+
+variables {s t}
+
+@[to_additive]
+lemma mul_salem_spencer.mono (h : t ⊆ s) (hs : mul_salem_spencer s) : mul_salem_spencer t :=
+λ a b c ha hb hc, hs (h ha) (h hb) (h hc)
+
+@[to_additive, simp]
```

Comment:
```suggestion
@[simp, to_additive]
```
And below where appropriate. `mul_salem_spencer_singleton` would be nice too, and maybe a `_pair` version?